Refine the heuristics for "substr: start index ... is past end of string" so
the warning is only shown if the third argument is set or if the start index is
more than 1 character beyond the end of a string.

--- builtin.c	2026-07-09 11:47:00.000000000 -0500
+++ builtin.c	2026-10-28 13:57:15.027376421 -0500
@@ -705,9 +705,7 @@
 
 	if (nargs == 3) {
 		if (! (d_length >= 1)) {
-			if (do_lint == DO_LINT_ALL)
-				lintwarn(_("substr: length %g is not >= 1"), d_length);
-			else if (do_lint == DO_LINT_INVALID && ! (d_length >= 0))
+			if (do_lint == DO_LINT_ALL && ! (d_length >= 0))
 				lintwarn(_("substr: length %g is not >= 0"), d_length);
 			DEREF(t1);
 			/*
@@ -780,7 +778,7 @@
 		src_len = t1->stlen;
 
 	if (indx >= src_len) {
-		if (do_lint)
+		if (do_lint && (nargs == 3 || indx > src_len))
 			lintwarn(_("substr: start index %g is past end of string"),
 				d_index);
 		DEREF(t1);
